sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -71,6 +71,26 @@ class SQLQuerySuite extends QueryTest with BeforeAndAfterAll {
     )
   }
 
+  test("SPARK-7158 collect and take return different results") {
+    import java.util.UUID
+    import org.apache.spark.sql.types._
+    val rdd = sparkContext.parallelize(List(1,2,3), 2)
+    val schema = StructType(List(StructField("index",IntegerType,true)))
+    val df = createDataFrame(rdd.map(p => Row(p)), schema)
+    def id:() => String = () => {UUID.randomUUID().toString()}
+    def square:Int => Int = (x: Int) => {x * x}
+    //expect the ID to have materialized at this point
+    val dfWithId = df.withColumn("id",callUDF(id, StringType)).cache()
+    val d1 = dfWithId.collect()
+
+    // ID should not be changed, as we will take the data from the cache directly
```

Comment:
this test could be simpler... no need for the additional dfWithIdAndSquare column or the square function...
if we were not reading from cache, two consecutive invocations of dfWithId.collect will yield different IDs.
Makes the test simpler.